CRM/Twingle/Profile.php

```diff
@@ -177,21 +193,137 @@ public function getTransactionID($twingle_id) {
    * Verifies whether the profile is valid (i.e. consistent and not colliding
    * with other profiles).
    *
-   * @throws Exception
+   * @throws \CRM_Twingle_Exceptions_ProfileValidationError
+   * @throws \Civi\Core\Exception\DBQueryException
    *   When the profile could not be successfully validated.
    */
-  public function verifyProfile() {
-    // TODO: check
-    //  data of this profile consistent?
-    //  conflicts with other profiles?
+  public function validate() {
+
+    // Name cannot be empty
+    if (empty($this->getName())) {
+      throw new ProfileValidationError(
+        'name',
+        E::ts('Profile name cannot be empty.'),
+        ProfileValidationError::ERROR_CODE_PROFILE_VALIDATION_FAILED
+      );
+    }
+
+    // Restrict profile names to alphanumeric characters, space and the underscore.
+    $contains_illegal_characters = preg_match("/[^A-Za-z0-9_\s]/", $this->getName());
+    if ($contains_illegal_characters) {
+      throw new ProfileValidationError(
+        'name',
+        E::ts('Only alphanumeric characters, space and the underscore (_) are allowed for profile names.'),
+        ProfileValidationError::ERROR_CODE_PROFILE_VALIDATION_FAILED
+      );
+    }
+
+    // Check if profile name is already used for other profile
+    $profile_name_duplicates = array_filter(
+      CRM_Twingle_Profile::getProfiles(), function($profile) {
+      return $profile->getName() == $this->getName() && $this->getId() != $profile->getId();
+    });
+    if (!empty($profile_name_duplicates)) {
+      throw new ProfileValidationError(
+        'name',
+        E::ts("A profile with the name '%1' already exists.", [1 => $this->getName()]),
+        ProfileValidationError::ERROR_CODE_PROFILE_VALIDATION_FAILED
+      );
+    }
+
+    // Check if project_id is already used in other profile
+    $profiles = $this::getProfiles();
+    foreach ($profiles as $profile) {
+      if ($profile->getId() == $this->getId() || $profile->is_default()) continue;
+      $project_ids = $this->getProjectIds();
+      $id_duplicates = array_intersect($profile->getProjectIds(), $project_ids);
+      if (!empty($id_duplicates)) {
+        throw new ProfileValidationError(
+          'selector',
+          E::ts(
+            "Project ID(s) [%1] already used in profile '%2'.",
+            [
+              1 => implode(", ", $id_duplicates),
+              2 => $profile->getName()
+            ]
+          ),
+          ProfileValidationError::ERROR_CODE_PROFILE_VALIDATION_FAILED
+        );
+      }
+    }
```

Comment:
I'd say having profiles with the same `selector` should not be prohibited per se, as you might want to test different configurations with the same project ID. I'm not sure, however, what happens in that case, as profiles can't be deactivated yet and we'd get multiple matches.

Reply:
From my user perspective, I would prefer to be warned when I use a selector in multiple profiles. Otherwise it could lead to unexpected behavior.

Also, the selectors are not very easy for human eyes to tell apart. In this case, the abstract requirement of being able to test multiple profiles with one selector (with an uncertain outcome of which profile is then applied) is something I would put behind improving the usability..

Reply:
I tend to agree but currently, using the same selector in multiple profiles has been possible and we'd make those configurations invalid with that change. What about not making it impossible but have the form issue a warning message only and still let the user save?

Reply:
Okay, I have added a new `error_code` called `ERROR_CODE_PROFILE_VALIDATION_WARNING` to `ProfileValidationError`. I use this code to decide if an error should be displayed in the form or just be a warning.

![Bildschirmfoto vom 2023-08-17 12-50-09](https://github.com/systopia/de.systopia.twingle/assets/24728186/d63562d5-75be-412e-9675-ede1468fdc3f)

_(Please don't be surprised to see the TwingleShop integration on the screenshot. Due to the dependencies between the branches that are not met, I had to test this on my latest branch)._